frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/Model.java

```diff
@@ -116,19 +122,49 @@ public void addFirst(Job job) {
         jobsDb.get(DEFAULT_DATA_QUEUE).addFirst(job);
     }
 
-    public Job nextJob(String threadId) throws InterruptedException {
-        return nextJob(threadId, Long.MAX_VALUE);
-    }
+    public void pollBatch(String threadId, long waitTime, Map<String, Job> jobsRepo)
+            throws InterruptedException {
+        if (jobsRepo == null || threadId == null) {
+            logger.error("Invalid input given provided");
+            throw new IllegalArgumentException();
+        }
+
+        if (!jobsRepo.isEmpty()) {
+            logger.error("The jobs repo provided contains stale jobs. Clear them!!");
+            throw new IllegalArgumentException();
+        }
 
-    public Job nextJob(String threadId, long timeoutMillis) throws InterruptedException {
-        LinkedBlockingDeque<Job> jobs = jobsDb.get(threadId);
-        if (jobs != null && !jobs.isEmpty()) {
-            Job j = jobs.poll();
+        LinkedBlockingDeque<Job> jobsQueue = jobsDb.get(threadId);
+        if (jobsQueue != null && !jobsQueue.isEmpty()) {
+            Job j = jobsQueue.poll(waitTime, TimeUnit.MILLISECONDS);
             if (j != null) {
-                return j;
+                jobsRepo.put(j.getJobId(), j);
+                return;
+            }
+        }
+
+        synchronized (this) {
```

Comment:
Can you explain why you're taking lock on the entire Model object itself ?

Reply:
There are two kind of queues,
1. Control queue, which is per threadId. This doens't need any synchronization as each thread has its own queue and they can pull from it without any locks.
2. Data queue: This is global and shared across all threads under a model for MMS. This requries global sync because at any given time only one thread should be accessing the batch of elements.